### PR TITLE
Add gitlab config for GH200 nodes

### PR DESCRIPTION
**Pull request by edopao**
Update the Gitlab configuration for CSCS Ext-CI:
- Add build and test configuration for Grace-Hopper nodes.
- Change test execution to use Firecrest for authentication, in order to mount project filesystem containing the test data.
- The pipeline on Grace-Hopper nodes is disabled, by default, and can be triggered manually.
- Include docker build arguments in the checksum DOCKER_TAG, in addition to the checksum on the docker file, in order to rebuild the base image if any argument is modified.

### CODE REVIEW

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
launch jenkins spack

**Comment by edopao** [comment]
cscs-ci run benchmark

**Comment by edopao** [comment]
cscs-ci run benchmark

**Comment by edopao** [comment]
cscs-ci run benchmark

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
launch jenkins spack

**Comment by edopao** [comment]
cscs-ci run dace

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

**Comment by edopao** [comment]
cscs-ci run default

File: model/tox.ini [hunk 1]
```diff
@@ -9,14 +9,17 @@ skipsdist = true
...
+    NUM_PROCESSES
     PIP_USER
     PYTHONUSERBASE
     TEST_DATA_PATH
 deps =
     -r {toxinidir}/requirements-dev.txt
 commands =
     pip list
-    -pytest -v -s -n auto --cache-clear --cov --cov-reset --doctest-modules atmosphere/dycore/src atmosphere/diffusion/src atmosphere/advection/src common/src driver/src
+    -pytest -v -s -n {env:NUM_PROCESSES:1} --cache-clear --cov --cov-reset --doctest-modules atmosphere/dycore/src atmosphere/diffusion/src atmosphere/advection/src common/src driver/src
```
Comment (samkellerhals) [inline model/tox.ini:22]: Is there a particular reason why you limit the num processes to 16 here?

File: ci/docker/base.Dockerfile [hunk 2]
```diff
@@ -76,4 +80,5 @@ RUN pyenv update && \
 
 ENV PATH="/root/.pyenv/shims:${PATH}"
 
-RUN pip install --upgrade pip setuptools wheel tox clang-format cupy-cuda11x
+ARG CUPY_PACKAGE=cupy-cuda11x
```
Comment (samkellerhals) [inline ci/docker/base.Dockerfile:83]: Is cupy-cuda11x the default, and then overriden by the CUPY_PACKAGE env var in the case in which it runs on gh, making it install cupy-cuda12x?

**Review by samkellerhals — APPROVED** [review]
Overall looks very nice, I just have a few questions for my own understanding.

File: ci/base.yml [hunk 3]
```diff
@@ -3,47 +3,68 @@ include:
...
     reports:
       dotenv: build.env
   variables:
     DOCKERFILE: ci/docker/base.Dockerfile
+    # change to 'always' if you want to rebuild, even if target tag exists already (if-not-exists is the default, i.e. we could also skip the variable)
     CSCS_REBUILD_POLICY: if-not-exists
-    DOCKER_BUILD_ARGS: '["PYVERSION=$PYVERSION", "CI_PROJECT_DIR=$CI_PROJECT_DIR"]'
-    <<: *py310
+    DOCKER_BUILD_ARGS: '["ARCH=$ARCH", "BASE_IMAGE=$BASE_IMAGE", "HPC_SDK_VERSION=$HPC_SDK_VERSION", "HPC_SDK_NAME=$HPC_SDK_NAME", "CUPY_PACKAGE=$CUPY_PACKAGE", "PYVERSION=$PYVERSION", "CI_PROJECT_DIR=$CI_PROJECT_DIR"]'
+build_baseimage_x86_64:
```
Comment (samkellerhals) [inline ci/base.yml:30]: Once daint is decomissioned will we still be able to run tests for a machine with x86_64 architecture?

Comment (edopao) [reply ci/base.yml:30]: I know there is a vCluster with AMD CPUs (x86_64) but I don't know if there are vCluster with x86_64 + GPU nodes.

File: ci/docker/base.Dockerfile (hunk 2 above)
Comment (edopao) [reply ci/docker/base.Dockerfile:83]: Correct. I thought I could leave some default value for all arguments.

File: model/tox.ini (hunk 1 above)
Comment (edopao) [reply model/tox.ini:22]: Yes. If we leave `auto` pytest will spawn 244 threads on the gh200 cpu, one for each core. In gt4py we have observed 2 kinds of issues with high parallelism: each test is running cmake and building a program from source files, and many builds in parallel could cause the error that too many files are open at the same time; it seems that the cuda MPS server was hanging sometimes. So I apply in icon4py the same restriction we applied in gt4py.

File: ci/base.yml [hunk 4]
```diff
@@ -55,11 +76,26 @@ build_image:
...
+    HPC_SDK_PATH: "/opt/nvidia/hpc_sdk/Linux_${ARCH}/22.11"
+.test_template_aarch64:
+  extends: [.container-runner-todi-gh200, .test_template]
+  needs: [build_image_aarch64]
+  variables:
+    CSCS_ADDITIONAL_MOUNTS: '["/store/migration/project/d121/icon4py/ci/testdata:$TEST_DATA_PATH"]'
+    HPC_SDK_PATH: "/opt/nvidia/hpc_sdk/Linux_${ARCH}/24.5"
+    # Grace-Hopper gpu architecture is not enabled by default in CUDA build
+    CUDAARCHS: "90"
+    # Limit test parallelism to avoid "OSError: too many open files" in the gt4py build stage.
```
Comment (edopao) [inline ci/base.yml:98]: This code comment addresses one of your questions.

File: ci/docker/base.Dockerfile (hunk 2 above)
Comment (edopao) [reply ci/docker/base.Dockerfile:83]: To be more precise, `CUPY_PACKAGE` is a docker build argument with default value `cupy-cuda11x`. When we build the base image for gh200, we set the build argument to `cupy-cuda12x`.

**Pull request merged**